docs(test): documentation for running GUI tests locally (#964)
opencloud-eu/desktop: test/gui/README.md

# Desktop Client GUI Testing

The OpenCloud desktop GUI tests use the following tools:

- **Behave** – Executes the GUI test scenarios.
- **Appium** – Drives the test automation by sending WebDriver commands.
- **PyAutoGUI** – Performs mouse and keyboard interactions that are not exposed through the accessibility API.
- **selenium-webdriver-at-spi** – A WebDriver implementation for Appium that uses the Linux AT-SPI accessibility API to automate desktop applications.

# Running GUI Tests

> ## Linux (Ubuntu)

Reviewer: let's be specific
```suggestion
## Linux (Ubuntu 24.04)
```

This guide explains how to build the OpenCloud desktop client and run the GUI test suite on Linux.

### Prerequisites

Install the following tools before continuing:

- Git
- Python 3
- Docker and Docker Compose
- CMake
- Ninja

**Install system dependencies**

```bash
sudo apt update

sudo apt install \
    xclip \
    python3-pyatspi \
    libatspi2.0-dev \
    libdbus-1-dev \
    libfuse3-dev \
    libfuse-dev
```

### 1. Install `uv` (Python environment and dependency manager)

Install [uv](https://astral.sh/uv) using the official installer:

```bash
curl -LsSf https://astral.sh/uv/install.sh | sh
```

Verify the installation:

```bash
uv --version
```

> [!NOTE]
> If `uv` is not found after installation, ensure that `~/.local/bin` is in your `PATH`.

### 2. Build the Desktop Client

> 1. **Install the required build dependencies using** [openbuild](https://github.com/opencloud-eu/openbuild)

Reviewer: ```suggestion
1. Install the required build dependencies using [openbuild](https://github.com/opencloud-eu/openbuild)
```

   ```bash
   python3 ./openbuild.py --branch main --target linux-gcc-x86_64 -- --install-deps opencloud-desktop
   ```

> [!IMPORTANT]
> Verify that Qt was built with accessibility support enabled:
>
> ```bash
> strings main/linux-gcc-x86_64/lib/libQt6Gui.so | grep -i QSpiAccessibleBridge
> ```
>

> > The command should print:

Reviewer: ```suggestion
> The output should contain:
```

>
> ```text
> QSpiAccessibleBridge
> ```

> 2. **Build the desktop client**

Reviewer: ```suggestion
2. Build the desktop client
```

Navigate to the desktop repository and build the client:

```bash
mkdir build
cd build

cmake \
    -DCMAKE_PREFIX_PATH=<path-to-openbuild>/main/linux-gcc-x86_64 \
    -GNinja \
    -DBUILD_TESTING=OFF \
    -DCMAKE_BUILD_TYPE=Debug \
    -DVIRTUAL_FILE_SYSTEM_PLUGINS=off \
    -S ..

ninja
```

### 3. Install the Test Dependencies

```bash
cd <desktop-repo-root>/test/gui
make install
```

### 4. Start the OpenCloud Server

Create a `compose.yaml` file with the following configuration:

```yaml
services:
   opencloud:
      image: opencloudeu/opencloud:latest
      entrypoint: /bin/sh
      command: ['-c', 'opencloud init || true && opencloud server']
      environment:
         OC_LOG_LEVEL: 'error'
         OC_INSECURE: true
         OC_URL: 'https://localhost:9200'
         IDM_ADMIN_PASSWORD: 'admin'
         PROXY_ENABLE_BASIC_AUTH: true
      ports:
         - 9200:9200
```

Start the server:

```bash
docker compose up
```

> [!NOTE]
> - Close any running OpenCloud desktop client before running the GUI tests.
> - Keep this terminal running while executing the GUI tests.

> 

Reviewer: ```suggestion
```

### 5. Start the AT-SPI WebDriver

Run the following commands to start the webdriver server:

```bash
cd <desktop-repo-root>/test/gui
source .venv/bin/activate
bash woodpecker/run_atspi_webdriver.sh
```

### 6. Configure the test environment

Copy `config.sample.ini` to `config.ini`, then update the required values.

```ini
[DEFAULT]
APP_PATH=<full-path-to>/opencloud # desktop app path
BACKEND_HOST=<opencloud-server-url>
```

### 7. Run the GUI Tests

> 

Reviewer: ```suggestion

> [!NOTE]
> - Close any running OpenCloud desktop client before running the GUI tests.

```

# Desktop Client GUI Testing

The OpenCloud desktop GUI tests use the following tools:

- **Behave** – Executes the GUI test scenarios.
- **Appium** with the **AT-SPI WebDriver** – Automates interactions with the OpenCloud desktop application on Linux.

> - **uv** – Manages the Python environment and project dependencies.

Reviewer: uv is not a test framework
```suggestion
```

> - **uv** – Manages the Python environment and project dependencies.

Reviewer: please also include, [selenium-webdriver-at-spi](https://invent.kde.org/sdk/selenium-webdriver-at-spi) and `PyAutoGUI`

# Running GUI Tests

## Linux (Ubuntu)

This guide explains how to build the OpenCloud desktop client and run the GUI test suite on Linux.

## Prerequisites

Install the following tools before continuing:

- Git
- Python 3
- Docker and Docker Compose
- CMake
- Ninja

## 1. Install Dependencies

> ### Install `uv` (Python environment and dependency manager)

Reviewer: headings level
```
## Linux (Ubuntu) -> lvl2
### Prerequisites -> lvl3
### Install Dependencies -> lvl3
```

Install [uv](https://astral.sh/uv) using the official installer:

```bash
curl -LsSf https://astral.sh/uv/install.sh | sh
```

Verify the installation:

```bash
uv --version
```

> [!NOTE]
> If `uv` is not found after installation, ensure that `~/.local/bin` is in your `PATH`.

### Install system dependencies

```bash
sudo apt update

sudo apt install \
    xclip \
    python3-pyatspi \
    libatspi2.0-dev \
    libdbus-1-dev \
    libfuse3-dev \

>     libfuse-dev

Reviewer: these should be in the prerequisites section

```

## 2. Build the Desktop Client

1. **Install the required build dependencies using** [openbuild](https://github.com/opencloud-eu/openbuild)

   ```bash
   python3 ./openbuild.py --branch main --target linux-gcc-x86_64 -- --install-deps opencloud-desktop
   ```

> [!IMPORTANT]
> Verify that Qt was built with accessibility support enabled:
>
> ```bash
> strings main/linux-gcc-x86_64/lib/libQt6Gui.so | grep -i QSpiAccessibleBridge
> ```
>
> The command should print:
>
> ```text
> QSpiAccessibleBridge
> ```

2. **Build the desktop client**

Navigate to the desktop repository and build the client:

```bash
mkdir build
cd build

cmake \
    -DCMAKE_PREFIX_PATH=<path-to-openbuild>/main/linux-gcc-x86_64 \
    -GNinja \
    -DBUILD_TESTING=OFF \
    -DCMAKE_BUILD_TYPE=Debug \
    -DVIRTUAL_FILE_SYSTEM_PLUGINS=off \
    -S ..

ninja
```

> ## 3. Install the project dependencies

Reviewer: maintain the heading case format: https://capitalizemytitle.com/

```suggestion
## 3. Install the Test Dependencies
```

```bash
cd <desktop-repo-root>/test/gui
make install
```

## 4. Start the OpenCloud Server

Create a `compose.yaml` file with the following configuration:

```
#compose.yaml
   services:
     opencloud:
      image: opencloudeu/opencloud:latest
      entrypoint: /bin/sh

>       command: ['-c', 'opencloud init || true && opencloud server']

Reviewer: also check the config indentation
```suggestion
```yaml
services:
  opencloud:
    image: opencloudeu/opencloud:latest
    entrypoint: /bin/sh
    command: ['-c', 'opencloud init || true && opencloud server']
```

      environment:
         OC_LOG_LEVEL: 'error'
         OC_INSECURE: true
         OC_URL: 'https://localhost:9200'
         IDM_ADMIN_PASSWORD: 'admin'
         PROXY_ENABLE_BASIC_AUTH: true
      ports:
         - 9200:9200
```

Start the server:

```bash
docker compose up
```

> [!NOTE]
> - Close any running OpenCloud desktop client before running the GUI tests.
> - Keep this terminal running while executing the GUI tests.

## 5. Start the AT-SPI WebDriver

> Open a new terminal and start the AT-SPI WebDriver:

Reviewer: ```suggestion
Run the following commands to start the webdriver server:
```

```bash
cd <desktop-repo-root>/test/gui
source .venv/bin/activate
bash woodpecker/run_atspi_webdriver.sh
```

## 6. Configure the test environment

Copy `config.sample.ini` to `config.ini`, then update the required values.

```ini
[DEFAULT]
APP_PATH=<full-path-to>/opencloud # desktop app path
BACKEND_HOST=<opencloud-server-url>
```

## 7. Run the GUI Tests

Run a test scenario:

```bash
cd <desktop-repo-root>/test/gui
uv run behave features/add-account/account.feature
```

## Test Reports

Following test reports are generated in `test/gui/reports/` path:

- **`report.log`** – Contains the detailed execution log.

> - **`report.html`** – An HTML report summarizing the test results.

Reviewer: ```suggestion
- **`report.html`** – An HTML report summarizing the test results.
- **`screenshots/`** – ...
- **`screenrecordings/`** – ...
```

Screenshot and screen recording of the test execution can be enabled using the following:

```ini
# test/gui/config.ini
CI=true
RECORD_VIDEO_ON_FAILURE=true
```

Or while running the test:

```bash
CI=true \
RECORD_VIDEO_ON_FAILURE=true \
uv run behave features/add-account/account.feature
```

> [!NOTE]
> When a test fails, a screen recording is automatically saved in:
>
> ```text
> test/gui/reports/recordings/

> > ```

Reviewer: ```suggestion
```

# Running GUI Tests Locally on Linux

> This guide explains how to build the OpenCloud desktop client and run the GUI test suite on Linux.

Reviewer: ```suggestion
# Desktop Client GUI Testing
<some words here about the test tools and framework used>

# Running GUI Tests

## Linux (Ubuntu)

This guide explains how to build the OpenCloud desktop client and run the GUI test suite on Linux.
```

## Prerequisites

Install the following tools before continuing:

- Git
- Python 3
- Docker and Docker Compose
- CMake
- Ninja

## 1. Install Dependencies

### Install `uv` (Python environment and dependency manager)

Install [uv](https://astral.sh/uv) using the official installer:

```bash

> curl -LsSf https://astral.sh/uv/install.sh | sh

Reviewer 2: Does it really have to be `uv`? pip is available on most systems anyway.

Also, please consider to install software from linux distros packages rather than curl'ing from a website. 

Maybe add something "Install uv from your distro package management or, if it is not available from curl..."

Reviewer: We used uv as it is faster than pip and helps in CI. But yeah, maybe we should allow using pip as well for the local environment.

Reviewer: we will check if can make a system for two different package managers

```

Verify the installation:

```bash
uv --version
```

> [!NOTE]
> If `uv` is not found after installation, ensure that `~/.local/bin` is in your `PATH`.

### Install system dependencies

```bash
sudo apt update

sudo apt install \
    xclip \
    python3-pyatspi \
    libatspi2.0-dev \
    libdbus-1-dev \
    libfuse3-dev \
    libfuse-dev

> ```

Reviewer 2: Please add on which distro that was tested, probably Ubuntu ?

Author: Tested on Ubuntu 24

## 2. Build the Desktop Client

### i. Build from Source

1. **Clone the openbuild repository** 

   ```bash
   git clone https://github.com/opencloud-eu/openbuild.git
   cd openbuild
   ```

2. **Install the required build dependencies**

   ```bash
   python3 ./openbuild.py --branch main --target linux-gcc-x86_64 -- --install-deps opencloud-desktop
   ```

> [!IMPORTANT]
> Verify that Qt was built with accessibility support enabled:
>
> ```bash
> strings ~/CraftRoot/lib/libQt6Gui.so | grep -i QSpiAccessibleBridge
> ```
>
> The command should print:
>
> ```text
> QSpiAccessibleBridge
> ```

3. **Clone the desktop client repository**

   ```bash
   cd ..
   git clone git@github.com:opencloud-eu/desktop.git
   cd desktop
   ```

4. **Create a build directory**:

   ```bash
   mkdir build
   cd build
   ```

5. **Generate the build files**

   ```bash
   cmake \
       -DCMAKE_PREFIX_PATH=<path-to-openbuild>main/linux-gcc-x86_64 \
       -GNinja \
       -DBUILD_TESTING=OFF \
       -DCMAKE_BUILD_TYPE=Debug \
       -DVIRTUAL_FILE_SYSTEM_PLUGINS=off \
       -S ..
   ```

> 6. **Build the desktop client:** `ninja`

Reviewer: let's not make the build process this long
```suggestion
### Build Desktop Client

1. Install the required build dependencies using [openbuild](https://github.com/opencloud-eu/openbuild)

    ```bash
    python3 ./openbuild.py --branch main --target linux-gcc-x86_64 -- --install-deps opencloud-desktop
    ```

> [!IMPORTANT]
> Verify that the installed Qt has accessibility feature:
>
> ```bash
> strings main/linux-gcc-x86_64/lib/libQt6Gui.so | grep -i QSpiAccessibleBridge
> ```

2. Go to the client directory and build it
    ```bash
    mkdir build && cd build
   
    cmake \
      -DCMAKE_PREFIX_PATH=<path-to-openbuild>main/linux-gcc-x86_64 \
      -GNinja \
      -DBUILD_TESTING=OFF \
      -DCMAKE_BUILD_TYPE=Debug \
      -DVIRTUAL_FILE_SYSTEM_PLUGINS=off \
      -S ..

    ninja
    ```
```

### ii. Using the AppImage

    1. Download the AppImage from the [releases page](https://github.com/opencloud-eu/desktop/releases)

    2. Extract the binary from the AppImage, using the command: `<path-to-appimage.AppImage> --appimage-extract`

    3. The binary will be extracted to `./squashfs-root/usr/bin` directory.
    

>     4. For more required environment variables, the app can be started through `./squashfs-root/AppRun.wrapped`.

Reviewer: let's not add appimage for now.
```suggestion
```

## 3. Clone the AT-SPI WebDriver

The AT-SPI WebDriver is a WebDriver server for Linux applications that uses the AT-SPI accessibility API, enabling Appium to automate the desktop client during GUI tests.
    
```bash
git clone https://invent.kde.org/sdk/selenium-webdriver-at-spi.git

> ```

Reviewer: this is not needed. we can use existing script:
```
bash test/gui/woodpecker/run_atspi_webdriver.sh
```

## 4. Set Up the Python Environment

Navigate to the GUI tests directory in the **desktop** repository (where the Behave tests reside):

```bash
cd <desktop-repo-root>/test/gui
```

Install the project dependencies:

```bash
uv sync

> ```

Reviewer: ```suggestion
## 4. Install the test dependencies:

    ```bash
    cd <desktop-repo-root>/test/gui
    make install
    ```
```

## 5. Start the OpenCloud Server

Start an OpenCloud server. For example, using the web repository:

```bash
git clone git@github.com:opencloud-eu/web.git
cd web
pnpm install
pnpm build
docker compose up opencloud
```

> [!IMPORTANT]
> Ensure the following environment variables are set before starting OpenCloud:
>
> - `IDM_ADMIN_PASSWORD=admin`
> - `PROXY_ENABLE_BASIC_AUTH=true`

> [!NOTE]
> - Close any running OpenCloud desktop client before running the GUI tests.

> > - Keep this terminal running while executing the GUI tests.

Reviewer: I would just provide a compose file
```suggestion
Add a docker compose file:
    ```
    #compose.yaml
    services:
      opencloud:
        image: opencloudeu/opencloud:latest
        entrypoint: /bin/sh
        command: ['-c', 'opencloud init || true && opencloud server']
        environment:
          OC_LOG_LEVEL: 'error'
          OC_INSECURE: true
          OC_URL: 'https://localhost:9200'
          IDM_ADMIN_PASSWORD: 'admin'
          PROXY_ENABLE_BASIC_AUTH: true
        ports:
          - 9200:9200
    ```

Run the server:
    ```bash
    docker compose up
    ```
```

## 6. Start the AT-SPI WebDriver

Open a new terminal and start the AT-SPI WebDriver.

**Option 1: Let `uv run` handle the environment automatically**

```bash
cd <desktop-repo-root>/test/gui

FLASK_APP=<path-to-selenium-webdriver-at-spi>/selenium-webdriver-at-spi.py \
FLASK_ENV=production \
uv run flask run \
    --port 4723 \
    --no-reload
```
    
**Option 2: Activate the virtual environment manually**

```bash
cd <desktop-repo-root>/test/gui
uv venv --system-site-packages
source .venv/bin/activate

FLASK_APP=<path-to-selenium-webdriver-at-spi>/selenium-webdriver-at-spi.py \
FLASK_ENV=production \
flask run \
    --port 4723 \
    --no-reload
```

> [!NOTE]
> If you get `ModuleNotFoundError: No module named 'pyatspi'`, the virtual environment is likely using a Python interpreter other than the system Python where `python3-pyatspi` is installed.
>
> Recreate the environment using the system Python:
>
> ```bash
> rm -rf .venv
> uv venv --python /usr/bin/python3 --system-site-packages
> ```
>
> You can verify the interpreter with:
>
> ```bash
> python --version
> python -c "import sys; print(sys.base_prefix)"
> ```
> 

> > Keep this terminal running while executing the GUI tests.

Reviewer: ```suggestion
## 6.  Start the AT-SPI WebDriver

    ```bash
    bash test/gui/woodpecker/run_atspi_webdriver.sh
    ````
```

> ## 7. Configure the Test Environment

Reviewer: ```suggestion
## 7. Configure the test environment
```

Copy `config.sample.ini` to `config.ini`, then update the required values.

```ini
[DEFAULT]
APP_PATH=<full-path-to>/opencloud # desktop app path
BACKEND_HOST=<opencloud-server-url>
```

## 8. Run the GUI Tests

**Option 1: Let `uv run` handle the environment automatically**

```bash
cd <desktop-repo-root>/test/gui
uv run behave features/add-account/account.feature
```

**Option 2: Activate the virtual environment manually**
    
```bash
cd <desktop-repo-root>/test/gui
source .venv/bin/activate
behave features/add-account/account.feature

> ```

Reviewer: ```suggestion
## 8. Run the GUI Tests

    ```bash
    cd <desktop-repo-root>/test/gui
    uv run behave features/add-account/account.feature
    ```
```

## 9. Reports and Screen Recordings

After the test execution completes, the following files are generated in `test/gui/reports/`:

- **`report.log`** – Contains the detailed execution log, including test steps, errors, and debugging information.
- **`report.html`** – An HTML report summarizing the test results, including passed and failed scenarios.

### Enable screen recording

Enable screen recording to record test execution and save recordings for failed test runs.

Edit `test/gui/config.ini`:

```ini
[DEFAULT]
RECORD_VIDEO_ON_FAILURE=true
 ```

**Alternatively, enable it using an environment variable**

   ```bash
   RECORD_VIDEO_ON_FAILURE=true uv run behave
   ```

> [!NOTE]
> When a test fails, a screen recording is automatically saved in:
>
> ```text
> test/gui/reports/recordings/

> > ```

Reviewer: ```suggestion
## Test Reports

Following test reports are generated in `test/gui/reports/` path:
- **`report.log`** – Contains the detailed execution log.
- **`report.html`** – An HTML report summarizing the test results.

Screenshot and screen recording of the test execution can be enabled using the following:

    ```
    # test/gui/config.ini
    CI=true
    RECORD_VIDEO_ON_FAILURE=true
    ```
Or while running the test:
    ```
    CI=true \
    RECORD_VIDEO_ON_FAILURE=true \
    uv run behave features/add-account/account.feature
    ```
```
